packages/firestore-types/index.d.ts

```diff
@@ -744,30 +763,39 @@ export class DocumentReference {
   /**
    * A reference to the Collection to which this DocumentReference belongs.
    */
-  readonly parent: CollectionReference;
+  readonly parent: CollectionReference<T>;
 
   /**
    * A string representing the path of the referenced document (relative
    * to the root of the database).
    */
   readonly path: string;
 
+  /**
+   * Firestore Converter to make the generic golem happy.
+   */
+  readonly converter: FirestoreConverter<T>;
+
   /**
    * Gets a `CollectionReference` instance that refers to the collection at
    * the specified path.
    *
    * @param collectionPath A slash-separated path to a collection.
    * @return The `CollectionReference` instance.
    */
-  collection(collectionPath: string): CollectionReference;
+  collection(collectionPath: string): CollectionReference<DocumentData>;
+  collection<C>(
+    collectionPath: string,
+    converter: FirestoreConverter<C>
+  ): CollectionReference<C>;
 
   /**
    * Returns true if this `DocumentReference` is equal to the provided one.
    *
    * @param other The `DocumentReference` to compare against.
    * @return true if this `DocumentReference` is equal to the provided one.
    */
-  isEqual(other: DocumentReference): boolean;
+  isEqual(other: DocumentReference<T>): boolean;
```

Comment:
How do you think we should handle DocumentReference<A>.isEqual(DocumentReference<B>) when both point to the same document? Does Typescipt allow us to do an equal check on the type of the Generic?

Reply:
Since we can't check for the type of Generic, Sebastian proposed adding an additional check to make sure the converters are the same, in addition to DocumentKey and underlying Firestore instance.